src/lib.rs

```diff
@@ -133,58 +169,67 @@ pub fn run_book(
-    if lazy {
-      let mem = Box::leak(hvmc::run::Heap::<true>::init(size)) as *mut _;
-      let net = hvmc::run::NetFields::<true>::new(unsafe { &*mem });
-      net.boot(hvmc::ast::name_to_val(entrypoint));
-      hvmc::run::Net::Lazy(hvmc::run::StaticNet { mem, net })
-    } else {
-      let mem = Box::leak(hvmc::run::Heap::<false>::init(size)) as *mut _;
-      let net = hvmc::run::NetFields::<false>::new(unsafe { &*mem });
-      net.boot(hvmc::ast::name_to_val(entrypoint));
-      hvmc::run::Net::Eager(hvmc::run::StaticNet { mem, net })
-    }
-  }
-}
-
 pub fn run_compiled(
-  book: &hvmc::ast::Book,
+  host: Arc<Mutex<Host>>,
   mem_size: usize,
   run_opts: RunOpts,
   hook: Option<impl FnMut(&Net)>,
   entrypoint: &str,
 ) -> (Net, RunStats) {
-  let runtime_book = book_to_runtime(book);
-  let root = &mut hvmc::run::Net::init(mem_size, run_opts.lazy_mode, entrypoint);
-
-  let start_time = Instant::now();
-
-  if let Some(mut hook) = hook {
-    expand(root, &runtime_book);
-    while !rdex(root).is_empty() {
-      hook(&net_from_runtime(root));
-      reduce(root, &runtime_book, 1);
-      expand(root, &runtime_book);
+  let heap = RtNet::<hvmc::run::Lazy>::init_heap(mem_size as usize);
+  let mut root = DynNet::new(&heap, run_opts.lazy_mode);
+  let max_rwts = run_opts.max_rewrites.map(|x| x.clamp(usize::MIN as u64, usize::MAX as u64) as usize);
+  // Expect won't be reached because there's
+  // a pass that checks this.
+  dispatch_dyn_net!(&mut root => {
+    root.boot(host.lock().unwrap().defs.get(entrypoint).expect("No main function."));
+
+    let start_time = Instant::now();
+
+    if let Some(mut hook) = hook {
+      root.expand();
+      while !root.rdex.is_empty() {
+        hook(&host.lock().unwrap().readback(&root));
+        root.reduce(1);
+        root.expand();
+      }
+    } else if let Some(mut max_rwts) = max_rwts {
+      if run_opts.lazy_mode {
+        panic!("Lazy mode does not yet support rewrite limit");
+      }
+      if !run_opts.single_core {
+        panic!("Parallel mode does not yet support rewrite limit");
+      }
+      root.expand();
+      while !root.rdex.is_empty() {
+        let old_rwts = root.rwts.total() as u64;
+        root.reduce(max_rwts);
+        let delta_rwts = root.rwts.total() as u64 - old_rwts;
+        if (max_rwts as u64) < delta_rwts {
+          eprintln!("Warning: Exceeded max rwts");
+          break;
+        }
+        max_rwts -= delta_rwts as usize;
+        root.expand();
+      }
+    } else if !run_opts.single_core {
+      root.parallel_normal();
+    } else {
+      root.normal();
     }
-  } else if run_opts.single_core {
-    root.normal(&runtime_book);
-  } else {
-    root.parallel_normal(&runtime_book);
-  }
-
-  let elapsed = start_time.elapsed().as_secs_f64();
+    let elapsed = start_time.elapsed().as_secs_f64();
+
+    let net = host.lock().unwrap().readback(&root);
+    // TODO I don't quite understand this code
```

Comment:
If I remember correctly, converting to Def is used here to make the inet dense again, so that we get the final size by just checking the length of the net.